stm32/machine_i2s: Add support for I2S on H7 MCUs (WIP)

dpgeorge opened the pull request:
Master TX is tested, in blocking and non-blocking IO mode.

To do:
- [x] test master RX mode
- [ ] test asyncio mode

(Note: this supports I2S using the SPI peripheral.  More sophisticated I2S support can be achieved using the SAI peripheral, but that's much more involved.)

dpgeorge committed:
stm32/machine_i2s: Add support for I2S on H7 MCUs.

Master TX is tested, in blocking and non-blocking IO mode.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/machine_i2s: Correctly set the DMA transfer size.

Signed-off-by: Damien George <damien@micropython.org>

miketeachman commented on ports/stm32/machine_i2s.c:
```diff
@@ -566,6 +565,21 @@ STATIC bool i2s_init(machine_i2s_obj_t *self) {
+        const pin_af_obj_t *pin_af = pin_find_af(self->sd, AF_FN_I2S, self->i2s_id);
+        if (pin_af != NULL) {
+            if ((pin_af->type == AF_PIN_TYPE_I2S_SDI
+                    && (self->mode == I2S_MODE_MASTER_RX || self->mode == I2S_MODE_SLAVE_RX))
```
What is the purpose of the slave mode macros, I2S_MODE_SLAVE_RX, I2S_MODE_SLAVE_TX?

dpgeorge replied:
I have another patch for preliminary I2S slave support on stm32, and this is needed for that to work.  I should put the slave PR up before this one...

miketeachman commented on ports/stm32/machine_i2s.c:
```diff
@@ -730,7 +746,12 @@ STATIC void machine_i2s_init_helper(machine_i2s_obj_t *self, size_t n_pos_args,
+        if (!(pin_af != NULL && (
+            pin_af->type == AF_PIN_TYPE_I2S_SD
+            #if defined(STM32H7)
+            || pin_af->type == AF_PIN_TYPE_I2S_SDO || pin_af->type == AF_PIN_TYPE_I2S_SDI
```
Can SDO and SDI pins work in either direction?  if not, it would be good to validate the SDO and SDI versus RX vs TX mode.

dpgeorge replied:
Yes, SDO and SDI can be swapped to work for either in or out.

miketeachman commented on ports/stm32/machine_i2s.c:
```diff
@@ -566,6 +565,21 @@ STATIC bool i2s_init(machine_i2s_obj_t *self) {
         GPIO_InitStructure.Alternate = (uint8_t)af->idx;
         HAL_GPIO_Init(self->sd->gpio, &GPIO_InitStructure);
+
+        #if defined(STM32H7)
```
Is there any other way to code this besides putting a processor macro into the file.  e.g use a feature support macro in mpconfigboard.h files?   Sometimes I find that processor macros seem to grow in micropython files, which tend to undermine code comprehension.  Not a big deal, and more of a question on my part to learn the best practice to follow in micropython.

dpgeorge replied:
I agree these things grow, but it's hard to make it clean.  The HAL is supposed to help there, and it does to some extent.

What's going on here is that the H7 (and maybe G4?) support SDI and SDO pins, and they can be swapped.  So instead of `#if defined(STM32H7)` it could be `#if MCU_SUPPORTS_SDI_SDO` (probably use a better name than that).  Does that sound reasonable?

miketeachman replied:
Thanks for explaining the issue.  Maybe just leave it as-is, and then consider refactoring if other processor variant(s) drive the use of additional macro logic?

miketeachman commented on ports/stm32/machine_i2s.c:
```diff
@@ -790,7 +811,9 @@ STATIC void machine_i2s_init_helper(machine_i2s_obj_t *self, size_t n_pos_args,
     init->MCLKOutput = I2S_MCLKOUTPUT_DISABLE;
     init->AudioFreq = args[ARG_rate].u_int;
     init->CPOL = I2S_CPOL_LOW;
+    #if !defined(STM32H7)
```
similar comment as above about processor macro vs feature macro

dpgeorge replied:
The H7 HAL just doesn't have this entry in the Init struct.  So somehow need to deal with a HAL that is not 100% cross-MCU compatible.

miketeachman commented on ports/stm32/machine_i2s.c:
```diff
@@ -584,13 +584,19 @@ STATIC bool i2s_init(machine_i2s_obj_t *self) {
     if (HAL_I2S_Init(&self->hi2s) == HAL_OK) {
         // Reset and initialize Tx and Rx DMA channels
+        uint32_t pm_size;
+        if (self->hi2s.Init.DataFormat == I2S_DATAFORMAT_16B) {
```
Is this a bug fix or DMA optimization for 32-bit audio?

dpgeorge replied:
This is a bug fix!  I'm surprised it works on other MCUs without this...

miketeachman replied:
Can you describe how the bug manifests?   I assume this bug was discovered during the H7 port.  Thanks.

dpgeorge replied:
When using 32-bit RX (which is always the case, it's always 32-bit for incoming data from a mic) the values read by the DMA were always 0.  And the DMA filled its buffer twice as fast.  This makes sense because the DMA was doing only 16-bit reads from the I2S data register, and two such reads are needed per 32-bit sample.

dpgeorge replied:
Also I'm pretty sure 32-bit TX was partially broken.  It did make sound on the speaker I used for testing, but it made "better" sound when I fixed this bug.

miketeachman replied:
Thanks.  That's clear now.  I'll take a closer look to see where I went wrong.  The I2S port to mimxrt is almost done.  I'll review the implementation to make sure the same bug doesn't show up on the mimxrt.

miketeachman replied:
After a more detailed look at the PR I do not see a bug in the original code:
1. The stm32 F4 and F7 documentation indicates that the I2S peripheral data width is a half-word. 
"DMA capability for transmission and reception (16-bit wide)" (32.3 of RM0385 Reference manual).   The I2S HAL docs discuss 16-bit transfers.
2. The Peripheral and Memory data width is specified as half-words in `dma.ca` in `dma_init_struct_i2s`.  
This structure data is used to initialize DMA in `dma_init_handle()` with `dma->Init = *dma_descr->init;`
3. Steps 1 and 2 would appear to correctly setup DMA for the F4 and F7 processor series.  

I am struggling to see a bug, at least for the F4 and F7 processors. The existing I2S implementation seems to follow
the manufacturers best practices.  

What information did you find to suggest a different approach to what is outlined above?  e.g.  using word size DMA.

dpgeorge replied:
I think you are right, I think the original code is correct for F4 and F7.

It's on H7 that the I2S data register (which now has separate TXDR and RXDR registers) behaves differently.

![h7i2s](https://user-images.githubusercontent.com/6187689/153703275-15799566-4cb7-4c0a-9c62-81c2fe337038.png)

miketeachman replied:
That makes sense to me now. 

Here is an implementation idea:  An alternative to overriding the dma data width sizes in the dma init struct (eg  dma_init_struct_i2s in dma.c) with the new function dma_init_with_size() is to modify the init structs in dma.c.

For example:
`const dma_descr_t dma_I2S_2_TX = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s };`

would change to:
`const dma_descr_t dma_I2S_2_TX_16 = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s_16 };`
and
`const dma_descr_t dma_I2S_2_TX_32 = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s_32 };`

There would be separate init structs for 16-bit vs 32-bit sample sizes.  These init structs would have macros for F4/F7/H7/etc to conditionally compile for uniqueness in the processor families, as it does now.

in machine_i2s.c, the code that looks like this:
```
        // configure DMA streams
        if (self->mode == I2S_MODE_MASTER_RX) {
            self->dma_descr_rx = &dma_I2S_1_RX;
        } else {
            self->dma_descr_tx = &dma_I2S_1_TX;
        }
```
would change to:

```
        // configure DMA streams
        if (self->mode == I2S_MODE_MASTER_TX && get_dma_bits(self->mode, self->bits) ==16  ) {
            self->dma_descr_rx = &dma_I2S_1_TX_16;
        } else if  (self->mode == I2S_MODE_MASTER_TX && get_dma_bits(self->mode, self->bits) ==32  ){
            self->dma_descr_tx = &dma_I2S_1_TX_32;
        } else if RX {
            self->dma_descr_rx = &dma_I2S_1_RX_32;    --- always 32 bits
        }
```
If this works, it has the advantage of keeping all the dma data width information in one place, dma.c -- which might aid in self-documenting the code.

miketeachman replied:
Another consideration for the H7 are the transformations that happen in machine_i2s.c code:
1.  transform for 32-bit sample TX:  https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/machine_i2s.c#L246
2. "cherry-picking" transform for RX: https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/machine_i2s.c#L308

You mentioned that master TX has been tested, so that should eliminate item 1 as a concern.  Hopefully the transform in item 2 will work as-is for the H7.

dpgeorge replied:
Yes I like your idea about having multiple DMA descriptor structs, for 16 and 32 bit transfers.   I'll try that out.

And will test again all combinations of TX, RX and bit size on H7.

dpgeorge replied:
I changed the DMA config so it uses separate 16 and 32 bit descriptor structs.

And I tested the byte transformations, they were wrong on the H7, and are now fixed (at least for 32-bit stereo).

miketeachman reviewed (commented):
Great that I2S is getting ported to another STM32 processor!   I have a few comments and questions about the proposed changes.

dpgeorge committed:
stm32/machine_i2s: Support I2S passive mode.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/machine_i2s: Set FullDuplexMode to disabled on F4.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/machine_i2s: Allow I2S.deinit to be called multiple times.

In particular, it is called by the constructor if the instance already
exists.  So if the previous instance was deinit'd then it will be deinit'd
a second time.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/multi_machine: Add a multi-test for machine.I2S.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
See #8451 for passive I2S support.  This PR would need to be rebased/reworked on top of that, and the multitest can then be used to test things.

dpgeorge committed:
stm32/dma: Provide 16 and 32 bit I2S DMA structs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Rebased on #8451.  Active and passive TX and RX now work on H7, but only 32-bit stereo has been tested.

dpgeorge committed:
stm32/machine_i2s: Fix 16-bit stereo i2s_frame_map.

Signed-off-by: Damien George <damien@micropython.org>

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.
